File: DynamoDB-SDK-Examples/python/WorkingWithQueries/query-scan-count.py
from __future__ import print_function  # Python 2/3 compatibility
import boto3
import json
import decimal
from boto3.dynamodb.conditions import Key, Attr
from botocore.exceptions import ClientError


# Helper class to convert a DynamoDB item to JSON.
class DecimalEncoder(json.JSONEncoder):
    def default(self, o):
        if isinstance(o, decimal.Decimal):
            return str(o)
        return super(DecimalEncoder, self).default(o)


dynamodb = boto3.resource('dynamodb', region_name='us-west-2')

table = dynamodb.Table('Music')

try:
    response = table.query(
        ExpressionAttributeNames={
            '#pk': 'Artist',
            '#sk': 'SongTitle',
        },
        ExpressionAttributeValues={
            ':pk': 'Michael Jackson',
            ':sk': 'A'
        },
        KeyConditionExpression='#pk = :pk AND begins_with(#sk, :sk)',
    )
except ClientError as e:
    print(e.response['Error']['Message'])
else:
    print('This query has scanned', response['ScannedCount'],
          'items and returned', response['Count'], 'items in total')

    if 'LastEvaluatedKey' in response:
        print('Not all items have been retrieved by this query. At least one another request is required to get all available items. The last evaluated key corresponds to',
              json.dumps(response['LastEvaluatedKey'], indent=4, cls=DecimalEncoder))

    print('Query response:', json.dumps(response, indent=4, cls=DecimalEncoder))

File: DynamoDB-SDK-Examples/node.js/WorkingWithQueries/query-scan-count.js
const AWS = require("aws-sdk");

const documentClient = new AWS.DynamoDB.DocumentClient({ region: "us-west-2" });

const query = async () => {
  const response = await documentClient
    .query({
      TableName: "Music",
      KeyConditionExpression: "#pk = :pk and begins_with(#sk, :sk)",
      ExpressionAttributeNames: {
        "#pk": "Artist",
        "#sk": "SongTitle",
      },
      ExpressionAttributeValues: {
        ":pk": "Michael Jackson",
        ":sk": "A",
      },
    })
    .promise();

  console.log(
    `This query has scanned ${response.ScannedCount} items and returned ${response.Count} items in total`
  );

  if (response.LastEvaluatedKey) {
    console.log(
      `Not all items have been retrieved by this query. At least one another request is required to get all available items. The last evaluated key corresponds to ${JSON.stringify(
        response.LastEvaluatedKey
      )}`
    );
  }

  console.log(`Query response: ${JSON.stringify(response, null, 2)}`);
};

query().catch((error) => console.error(JSON.stringify(error, null, 2)));
